### PR TITLE
Add ManagedMap::range.

### PR DESCRIPTION
**Pull request by ProgVal**
As mentioned here: https://github.com/m-labs/smoltcp/issues/209#issuecomment-388892966

### CODE REVIEW

**Comment by ProgVal** [comment]
Got it!

File: src/map.rs [hunk 1]
```diff
@@ -90,6 +96,103 @@ impl<T> Into<Option<T>> for RevOption<T> {
     }
 }
 
+pub enum Range<'a, 'b, K: 'a, V: 'a> where 'a: 'b {
```
Comment (whitequark) [inline src/map.rs]: Hmm, is there a reason the lifetimes have to be like this? Why not `K: 'b, V: 'b`?

File: src/map.rs [hunk 2]
```diff
@@ -90,6 +96,103 @@ impl<T> Into<Option<T>> for RevOption<T> {
...
+    /// Borrowed variant.
+    Borrowed(&'b [Option<(K, V)>], usize),
+    /// Owned variant, only available with the `std` or `alloc` feature enabled.
+    #[cfg(any(feature = "std", feature = "alloc"))]
+    Owned(BTreeRange<'b, K, V>),
+    _Phantom(&'a K),
```
Comment (whitequark) [inline src/map.rs]: This should, at least, be `#[doc(hidden)]`, and I usually go with `__Phantom` (two underscores).

File: src/map.rs [hunk 3]
```diff
@@ -90,6 +96,103 @@ impl<T> Into<Option<T>> for RevOption<T> {
...
+        else if middle > 0 && key!(slice[middle-1]) >= Some(key_begin) {
+            right = middle-1;
+        }
+        else {
+            left = middle;
+            right = middle;
+            break;
+        }
+    }
+    assert_eq!(left, right);
```
Comment (whitequark) [inline src/map.rs]: When will this assertion fail?

File: src/map.rs [hunk 4]
```diff
@@ -90,6 +96,103 @@ impl<T> Into<Option<T>> for RevOption<T> {
...
+        else if middle < slice.len() && key!(slice[middle]) < Some(key_end) {
+            left = middle+1;
+        }
+        else {
+            left = middle;
+            right = middle;
+            break;
+        }
+    }
+    assert_eq!(left, right);
```
Comment (whitequark) [inline src/map.rs]: When will this assertion fail?

File: src/map.rs [hunk 5]
```diff
@@ -90,6 +96,103 @@ impl<T> Into<Option<T>> for RevOption<T> {
...
+    // two loops here, because we need the left-most match (for begin) and
+    // the right-most match (for end), and the stdlib does not provide
+    // any of these guarantees.
+
+    // Find the beginning
+    while left < right {
+        let middle = left + (right-left)/2;
+        if key!(slice[middle]) < Some(key_begin) {
+            left = middle+1;
+        }
```
Comment (whitequark) [inline src/map.rs]: Nit: `} else` here and elsewhere.

File: src/map.rs [hunk 6]
```diff
@@ -155,6 +258,23 @@ impl<'a, K: Ord + 'a, V: 'a> ManagedMap<'a, K, V> {
         }
     }
 
+    pub fn range<'b, 'c, Q>(&'c self, key_begin: &Q, key_end: &Q) -> Range<'a, 'b, K, V>
+            where K: Borrow<Q>, Q: Ord + ?Sized, 'a: 'b, 'c: 'a
```
Comment (whitequark) [inline src/map.rs]: Is `'c` really necessary here?

Comment (ProgVal) [reply src/map.rs]: Yes, the `ManagedMap` has to outlive the `Range`'s references.

File: src/map.rs [hunk 7]
```diff
@@ -152,17 +149,13 @@ fn binary_search_by_key_range<K, V, Q>(slice: &[Option<(K, V)>], key_begin: &Q,
...
             right = middle-1;
-        }
-        else {
+        } else {
             left = middle;
-            right = middle;
             break;
         }
     }
-    assert_eq!(left, right);
```
[the remark this answers is not in the document]
Comment (ProgVal) [reply src/map.rs]: No, it can't fail. I used it while debugging

**Comment by whitequark** [comment]
Thanks. What do you think about removing rustc version detection entirely? I was thinking about using impl Trait in smoltcp and that would imply 1.26. I don't think we care very much about 1.25 or earlier.

**Comment by ProgVal** [comment]
Okay. I currently have only rust 1.25 on my computer, but I can fix that.

**Comment by ProgVal** [comment]
I wonder if I should add "shortcuts", to skip a dichotomy if `key_begin <= slice[0]` or `key_end > slice[slice.len()-1]`.

https://github.com/m-labs/smoltcp/issues/209 would be one use case, because we'll use `0.0.0.0/0` or `::/0` as the `key_begin`.

**Comment by ProgVal** [comment]
I just improved my implementation to have an interface similar to `BTreeMap::range`, and support more range types than "left included, right excluded". (we'll need to include right in smoltcp.)

It uses the nightly feature "collections_range", but it is [going to be stabilized soon](https://github.com/rust-lang/rust/issues/30877); and `ManagedMap` already depends on "slice_rotate" anyway.

**Comment by whitequark** [comment]
Thanks.

**Pull request merged**